src/vsr/replica.zig

```diff
@@ -5511,39 +5699,48 @@ pub fn ReplicaType(
             assert(header_head.op >= self.commit_min);
             assert(header_head.op >= self.commit_max);
+            assert(header_head.op <= self.op_checkpoint_trigger());
+            for (dvcs_all.constSlice()) |dvc| assert(header_head.op >= dvc.header.commit);
+
+            // When computing the new commit_max, we cannot simply rely on the fact that our own
+            // commit_min is attached to our own DVC in the quorum.
+            // Consider the case:
+            // 1. Start committing op=N…M.
+            // 2. Send `do_view_change` to self.
+            // 3. Finish committing op=N…M.
+            // 4. Remaining `do_view_change` messages arrive, completing the quorum.
+            // In this scenario, our own DVC's commit is `N-1`, but `commit_min=M`.
+            // Don't let the commit backtrack.
+            const commit_max = std.math.max(
+                self.commit_min,
+                DVCQuorum.commit_max(self.do_view_change_from_all_replicas),
+            );
+            assert(self.commit_min >=
+                self.do_view_change_from_all_replicas[self.replica].?.header.commit);
 
-            if (header_head.op > self.op_checkpoint_trigger()) {
-                // This replica is too far behind, i.e. the new `self.op` is too far ahead of the
-                // last checkpoint. If we wrap now, we overwrite un-checkpointed transfers in the WAL,
-                // precluding recovery.
-                //
-                // TODO State transfer. Currently this is unreachable because the
-                // primary won't checkpoint until all replicas are caught up.
-                stdx.unimplemented("state transfer");
-            }
+            self.set_op_and_commit_max(header_head.op, commit_max, "on_do_view_change");
+            assert(self.commit_max <= self.op_checkpoint_trigger());
+            assert(self.commit_max <= self.op);
 
-            self.set_op_and_commit_max(
-                header_head.op,
-                std.math.max(
-                    self.commit_max,
-                    std.math.max(
-                        // `set_op_and_commit_max()` expects the highest commit_max that we know of.
-                        // But DVCs include replica's `commit_min`, not `commit_max`.
-                        do_view_change_commit_min_max,
-                        // An op cannot be uncommitted if it is definitely outside the pipeline.
-                        // Use `do_view_change_op_head` instead of `replica.op` since the former is
-                        // about to become the new `replica.op`.
-                        header_head.op -| constants.pipeline_prepare_queue_max,
-                    ),
-                ),
-                "on_do_view_change",
-            );
             // "`replica.op` exists" invariant may be broken briefly between set_op_and_commit_max()
             // and replace_header().
             self.replace_header(&header_head);
```

Comment:
micro nit, but we may emphasize the extent of broken invariant a bit

```zig
            {
                // "`replica.op` exists" invariant may be broken briefly between set_op_and_commit_max()
                // and replace_header().

                self.set_op_and_commit_max(header_head.op, commit_max, "on_do_view_change");
                assert(self.commit_max <= self.op_checkpoint_trigger());
                assert(self.commit_max <= self.op);

                self.replace_header(&header_head);
                assert(self.journal.header_with_op(self.op) != null);
            }
```

Reply:
Good idea, and I can add a `maybe(self.journal.header_with_op(self.op) == null);` too.